core/src/main/java/com/yahoo/imapnio/async/response/ImapResponseMapper.java

```diff
@@ -443,18 +456,102 @@ private SearchResult parseToSearchResult(@Nonnull final IMAPResponse[] ir) throw
+            if (ir.length < 1) {
+                throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
+            }
+            final Response taggedResponse = ir[ir.length - 1];
+            taggedResponse.skipSpaces();
+            final byte bracket = taggedResponse.peekByte();
+            if (!taggedResponse.isOK() && bracket != (byte) L_BRACKET) { // BAD or NO without MODIFIED
+                throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
+            }
+            final List<IMAPResponse> imapResponses = new ArrayList<>(); // will always return a non-null array
+
+            MessageNumberSet[] modifiedMsgsets = null; // only one response will contain modified numbers
+            Long highestModSeq = null;
+
+            for (final IMAPResponse sr: ir) {
+                if (sr.keyEquals("FETCH")) {
+                    imapResponses.add(new FetchResponse(sr));
+                    continue;
+                }
+
+                sr.skipSpaces();
+                if (sr.readByte() == (byte) L_BRACKET) { // HIGHESTMODSEQ or MODIFIED
+                    final String responseCode = sr.readAtom();
+                    if (sr.isOK() && responseCode.equalsIgnoreCase("HIGHESTMODSEQ")) {
+                        highestModSeq = sr.readLong();
+                    } else if (responseCode.equalsIgnoreCase("MODIFIED")) {
+                        modifiedMsgsets = MessageNumberSet.buildMessageNumberSets(sr.readAtom());
+                    } else { // OK or NO with '[' but response code is not recognized.
+                        throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
                     }
+                } else { // OK, Vanished, or Expunge Responses
+                    imapResponses.add(sr);
+                }
+            }
+
+            return new StoreResult(highestModSeq, imapResponses, modifiedMsgsets);
+        }
+
+        /**
+         * Parses the responses from Fetch command and UID Fetch command to a @{code FetchResult} object.
+         *
+         * @param ir the list of responses from Fetch or UID Fetch command, the input responses array should contain the tagged/final one
+         * @return FetchResult object constructed based on the given IMAPResponse array,
+         * @throws ImapAsyncClientException when tagged response is not OK or given response length is 0
+         */
+        @Nonnull
+        private FetchResult parseToFetchResult(@Nonnull final IMAPResponse[] ir) throws ImapAsyncClientException, IOException, ProtocolException {
+            if (ir.length < 1) {
+                throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
+            }
+            final Response taggedResponse = ir[ir.length - 1];
+            if (!taggedResponse.isOK()) {
+                throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
```

Comment:
Throw this exception only when the result is BAD. Not for NO

Reply:
Currently, we use invalid input because it is a parser which we only care about the input. Server error exception should not be thrown here.